Please add com.github.ransome1.sleek

ransome1 opened the pull request:
Hi,

I'm Robin, author of sleek, which is a simple todo.txt manager written mostly in JS, HTML and CSS. It is an Electron app.

I hope the manifest meets Flathubs requirements. Please let me know if I did anything wrong.

The code is documented on Github: https://github.com/ransome1/sleek

It is also published on Snap Store: https://snapcraft.io/sleek

Best
Robin

bilelmoussaoui commented on org.flathub.sleek.yml:
```diff
@@ -0,0 +1,57 @@
+app-id: org.flathub.sleek
```
I don't think that's a proper app-id. Something like `com.github.ransome1.sleek` might work.

ransome1 replied:
changed it according to your suggestion

bilelmoussaoui commented on org.flathub.sleek.yml:
```diff
@@ -0,0 +1,57 @@
+app-id: org.flathub.sleek
+runtime: org.freedesktop.Platform
+runtime-version: '1.6'
```
Bump this to 20.08

ransome1 replied:
Adjusted it in another commit

bilelmoussaoui commented on org.flathub.sleek.yml:
```diff
@@ -0,0 +1,57 @@
+  - --socket=pulseaudio
+  - --share=network
+build-options:
+  cflags: -O2 -g
+  cxxflags: -O2 -g
```
```suggestion
```

LeandroStanger commented:
bot, build add org.flathub.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
@bilelmoussaoui , @LeandroStanger , do you guys see the error why build is failing? I only could think of this being the reason:
```
which: no python2.7 in (/app/bin:/usr/bin)
which: no python2 in (/app/bin:/usr/bin)
Please use Python 2.7
]2;flatpak-builder: Cleanup nodejsError: module nodejs: Child process exited with code 1
```
How would I solve this?

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
Does anybody have an idea, why the build process can't access npm?

`npm ERR! request to https://registry.npmjs.org/i18next failed, reason: getaddrinfo EAI_AGAIN registry.npmjs.org`

barthalion commented on com.github.ransome1.sleek.desktop:
```diff
@@ -0,0 +1,9 @@
+[Desktop Entry]
```
This should be part of your code repository, it's not specific to Flathub. Same about the icon file.

ransome1 replied:
mmhh, I checked other PR for flathub and they did it the same. But I can but those files on the source repo too, no problem.

barthalion replied:
Bump. Same about the icon file.

barthalion commented on com.github.ransome1.sleek.yml:
```diff
@@ -0,0 +1,68 @@
+runtime-version: '20.08'
+branch: stable
+sdk: org.freedesktop.Sdk
+base: io.atom.electron.BaseApp
```
```suggestion
base: org.electronjs.Electron2.BaseApp
```

barthalion commented on com.github.ransome1.sleek.yml:
```diff
@@ -0,0 +1,68 @@
+branch: stable
+sdk: org.freedesktop.Sdk
+base: io.atom.electron.BaseApp
+base-version: stable
```
```suggestion
base-version: 20.08```

barthalion commented on com.github.ransome1.sleek.yml:
```diff
@@ -0,0 +1,68 @@
+  #    - type: archive
+  #      url: https://nodejs.org/dist/v12.18.3/node-v12.18.3.tar.xz
+  #      sha256: 71158026579487422fd13cc2553b34cddb76519098aa6030faab52f88c6e0d0e
+  - name: sleek
```
Look at https://github.com/flatpak/flatpak-builder-tools/tree/master/node for examples how to build it offline on Flathub infra.

ransome1 replied:
I don't know what this means.

ransome1 replied:
Did you mean with "build it offline" to build the "generated-sources.json" offline? That is what I did in the first place, but using the "flatpak-npm-generator.py" to do so, which I now understand is outdated? I now used the "flatpak-node-generator.py" and will try building again.

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
@barthalion , same problem even with all the adjustments. During build process the connection to npmjs.org cannot be established. I can access the link on my computer and it looks healthy. Is it a network problem on Flathubs server side?

`npm ERR! request to https://registry.npmjs.org/i18next failed, reason: getaddrinfo EAI_AGAIN registry.npmjs.org`

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
@LeandroStanger , @barthalion , @bilelmoussaoui ,
 I did all I can do and will give up now. The problems persist. If I ask npm (or yarn, I tried both) to install the modules "online" it can't connect to npmjs.org, see: https://flathub.org/builds/#/builders/21/builds/2484

If I do it with the --offline flag and the updated "generated-sources.json" (using the python script here: https://github.com/flatpak/flatpak-builder-tools/tree/master/node) it complains about not being able to find cached sources for my dependencies, see: https://flathub.org/builds/#/builders/32/builds/33509

bilelmoussaoui commented:
Please have some patience and wait till someone either review your PR or guide you to fix your issue. Thanks

barthalion commented:
Hey @refi64, mind having a look when you have a minute? npm seems to insist on installing electron from Internet instead of re-using cached one.

```
> electron@11.0.3 postinstall /run/build/sleek/sleek/node_modules/electron
> node install.js

RequestError: getaddrinfo EAI_AGAIN github.com
    at ClientRequest.<anonymous> (/run/build/sleek/sleek/node_modules/got/source/request-as-event-emitter.js:178:14)
    at Object.onceWrapper (events.js:421:26)
    at ClientRequest.emit (events.js:326:22)
    at ClientRequest.origin.emit (/run/build/sleek/sleek/node_modules/@szmarczak/http-timer/source/index.js:37:11)
    at TLSSocket.socketErrorListener (_http_client.js:427:9)
    at TLSSocket.emit (events.js:314:20)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
```

bot, build com.github.ransome1.sleek

refi64 commented on com.github.ransome1.sleek.yml:
```diff
@@ -0,0 +1,48 @@
+    buildsystem: simple
+    build-commands:
+      - npm install --prefix=sleek --offline --cache=$FLATPAK_BUILDER_BUILDDIR/flatpak-node/npm-cache
+      - . flatpak-node/electron-builder-arch-args.sh; yarn run --offline dist -- $ELECTRON_BUILDER_ARCH_ARGS
```
I think this is the problem, this is using `npm` to download the dependencies but `yarn` to run the build. (This isn't really something supported outside Flatpak either afaik, you'll run into weird lockfile issues.) If the app uses yarn, follow the `yarn install` command in the flatpak-node-generator examples, and run it as `flatpak-node-generator yarn ...`.

barthalion replied:
Good catch; the electron failure happens on `npm install` step though. Not sure why, it's there in generated-sources.

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
@refi64 . Thanks for your suggestion. I changed the install command to yarn. The generated-sources.json was already generated according to your suggestion.

The next error now is `error Couldn't find a package.json file in "/run/build/sleek"`

I added a "cd /run/build/sleek" as I thought yarn tried to build from an empty directory. But that didn't prevent the error. So for some reason now the package.json from my repo isn't found in build directory.

barthalion commented:
@ransome1 You've got it wrong – as you are not using yarn, you should stick to npm. Instead of replacing `npm install` line, the `yarn run` one needs fixing. (It doesn't fix the problem though.)

ransome1 commented:
@barthalion . But in my yml I define yarn as a module:
```
  - name: yarn
    buildsystem: simple
    build-commands:
      - 'cp -a * /app'
    # Only used for building, so clean it up afterwards.
    cleanup:
      - '*'
    sources:
      - type: archive
        url: https://github.com/yarnpkg/yarn/releases/download/v1.19.1/yarn-v1.19.1.tar.gz
        sha256: 34293da6266f2aae9690d59c2d764056053ff7eebc56b80b8df05010c3da9343
```
So my guess is, that enables the build process to use yarn, no?

barthalion commented:
I don't think yarn uses the same lock file and cache location that npm does.

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

barthalion commented:
bot, build com.github.ransome1.sleek

barthalion commented:
bot, build com.github.ransome1.sleek

barthalion commented:
bot, build com.github.ransome1.sleek

barthalion commented:
bot, build com.github.ransome1.sleek

barthalion commented:
I'm getting somewhere, this time it's electron-builder trying to download electron for whatever reason:

```
  • electron-builder  version=22.9.1 os=5.9.11-200.fc33.x86_64
  • loaded configuration  file=package.json ("build" field)
  • writing effective config  file=dist/builder-effective-config.yaml
  • packaging       platform=linux arch=x64 electron=11.0.3 appOutDir=dist/linux-unpacked
  ⨯ Get "https://github.com/electron/electron/releases/download/v11.0.3/electron-v11.0.3-linux-x64.zip": dial tcp: lookup github.com on 127.0.0.1:53: read udp 127.0.0.1:59025->127.0.0.1:53: read: connection refused
github.com/develar/app-builder/pkg/download.(*Downloader).follow.func1
	/Volumes/data/Documents/app-builder/pkg/download/downloader.go:206
github.com/develar/app-builder/pkg/download.(*Downloader).follow
	/Volumes/data/Documents/app-builder/pkg/download/downloader.go:234
github.com/develar/app-builder/pkg/download.(*Downloader).DownloadNoRetry
	/Volumes/data/Documents/app-builder/pkg/download/downloader.go:128
github.com/develar/app-builder/pkg/download.(*Downloader).Download
	/Volumes/data/Documents/app-builder/pkg/download/downloader.go:112
github.com/develar/app-builder/pkg/electron.(*ElectronDownloader).doDownload
	/Volumes/data/Documents/app-builder/pkg/electron/electronDownloader.go:192
github.com/develar/app-builder/pkg/electron.(*ElectronDownloader).Download
	/Volumes/data/Documents/app-builder/pkg/electron/electronDownloader.go:177
github.com/develar/app-builder/pkg/electron.downloadElectron.func1.1
	/Volumes/data/Documents/app-builder/pkg/electron/electronDownloader.go:73
github.com/develar/app-builder/pkg/util.MapAsyncConcurrency.func2
	/Volumes/data/Documents/app-builder/pkg/util/async.go:68
runtime.goexit
	/usr/local/Cellar/go/1.14.5/libexec/src/runtime/asm_amd64.s:1373  
  ⨯ /run/build/sleek/sleek/node_modules/app-builder-bin/linux/x64/app-builder exited with code ERR_ELECTRON_BUILDER_CANNOT_EXECUTE  stackTrace=
```

barthalion commented:
bot, build com.github.ransome1.sleek

barthalion commented:
bot, build com.github.ransome1.sleek

barthalion commented:
Okay, here we are! @ransome1 your application needs metainfo and desktop files. You can have a look at https://github.com/flathub/flathub/pull/1987/files#diff-997cc22d8fcadbc30e4f6c2b1c4eefc2ce64ca40e708e3cd4d432bab05c6c085 for inspiration; https://www.freedesktop.org/software/appstream/metainfocreator/#/ can be used to generate the former. It won't be 100% correct for our needs but I'll help with fixing it when you have it ready.

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

hfiguiere commented on com.github.ransome1.sleek.yml:
```diff
@@ -0,0 +1,60 @@
+  - name: icon
+    buildsystem: simple
+    build-commands:
+      - install -Dm644 com.github.ransome1.sleek.png /run/build/icon/com.github.ransome1.sleek.png
```
you are installing in place which is why you have the build error.

The icon should be installed in `/app/share/icons/hicolor/128x128/apps/`.

ransome1 replied:
> you are installing in place which is why you have the build error.
> 
> The icon should be installed in `/app/share/icons/hicolor/128x128/apps/`.

I changed the path, but it didn't solve the problem, unfortunately.

hfiguiere replied:
it did, it's a different build failure.

ransome1 commented:
bot, build com.github.ransome1.sleek

hfiguiere commented on com.github.ransome1.sleek.appdata.xml:
```diff
@@ -0,0 +1,33 @@
+    <screenshot>
+      <image>https://raw.githubusercontent.com/ransome1/sleek/master/assets/screenshots/main_filter_dark.png</image>
+    </screenshot>
+  </screenshots>
```
This appdata doesn't validate:
```
• tag-missing           : <content_rating> required [use https://odrs.gnome.org/oars]
• tag-missing           : <release> required
• tag-missing           : <url> is not present
```

hfiguiere commented on com.github.ransome1.sleek.desktop:
```diff
@@ -0,0 +1,9 @@
+Name=com.github.ransome1.sleek
+Exec=/app/bin/run.sh %u
+Type=Application
+Icon=/app/share/applications/com.github.ransome1.sleek.png
```
`Icon` should just be the icon name, without extension. That's why the build is failing.

hfiguiere commented on com.github.ransome1.sleek.desktop:
```diff
@@ -0,0 +1,9 @@
+Exec=/app/bin/run.sh %u
+Type=Application
+Icon=/app/share/applications/com.github.ransome1.sleek.png
+Categories=Productivity
```
`Productivity` isn't a valid category. See https://specifications.freedesktop.org/menu-spec/latest/apa.html

ransome1 commented:
bot, build com.github.ransome1.sleek

hfiguiere commented on com.github.ransome1.sleek.appdata.xml:
```diff
@@ -8,6 +8,12 @@
+  <releases>
+    <release version="0.4.1" date="2020-12-13"/>
+  </releases>
+  <url type="github">https://github.com/ransome1/sleek</url>
```
this as an invalid type.
See https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-url

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
`(appstream-compose:2): As-CRITICAL **: 17:14:58.637: as_utils_find_icon_filename_full: assertion 'search != NULL' failed`

Mhh, back to the icon I guess.

hfiguiere commented:
> `(appstream-compose:2): As-CRITICAL **: 17:14:58.637: as_utils_find_icon_filename_full: assertion 'search != NULL' failed`
> 
> Mhh, back to the icon I guess.

I said:
> should just be the icon name, without extension

But you still try to put a path, which is invalid.

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
bot, build com.github.ransome1.sleek

ransome1 commented:
Thanks a lot for all your support guys. The build has been successful and after a couple of minor adjustments the app starts and works like a charm :)

hfiguiere commented on com.github.ransome1.sleek.yml:
```diff
@@ -11,6 +11,7 @@ finish-args:
   - --socket=x11
   - --share=network
   - --device=dri
+  - --filesystem=host
```
why "host" isn't "home" enough?

ransome1 replied:
Because home doesn't include external hdd or network shares. And the use case here is that a user is suppose to choose by himself, where he wants to save or open his todo.txt file from.

ransome1 replied:
in case this is a blocker out of security reasons, I changed it to "home".

ransome1 commented:
bot, build com.github.ransome1.sleek

barthalion commented on com.github.ransome1.sleek.appdata.xml:
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Ideally you should keep it in Sleek repo.

barthalion commented on com.github.ransome1.sleek.yml:
```diff
@@ -0,0 +1,60 @@
+        url: https://github.com/ransome1/sleek.git
+        commit: 1bcdc188a9f47eb08116fe532bd8e2ba076b7b04
+        dest: sleek
+      - generated-sources.json
```
FYI, this has been generated with `./flatpak-node-generator.py npm --xdg-layout package-lock.json`.

barthalion commented:
In general LGTM, I'll proceed with the merge. Please handle moving appdata/desktop/icon files to your repo when you have spare time.

/merge

ransome1 commented:
> In general LGTM, I'll proceed with the merge. Please handle moving appdata/desktop/icon files to your repo when you have spare time.
> 
> /merge

I tried that before, but didn't succeed on doing this as I don't know how to adjust the yml to work with the files from sleek repo. So I ended up putting the files in the Flathub repo.
